src/gui/widget_task.c

```diff
@@ -94,32 +218,405 @@ static void HandleRefreshStyle(LCUI_Widget w)
 	w->task.states[LCUI_WTASK_UPDATE_STYLE] = FALSE;
 }
 
-static void HandleUpdateStyle(LCUI_Widget w)
+static void Widget_OnUpdateStyle(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
 {
 	Widget_ExecUpdateStyle(w, FALSE);
 }
 
-static void HandleSetTitle(LCUI_Widget w)
+static void Widget_OnSetTitle(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
 {
 	Widget_PostSurfaceEvent(w, LCUI_WEVENT_TITLE, TRUE);
 }
 
-/** 处理主体刷新（标记主体区域为脏矩形，但不包括阴影区域） */
-static void HandleBody(LCUI_Widget w)
+static void Widget_OnRefreshBody(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
+{
+	if (ctx->diff.invalid_box < SV_BORDER_BOX) {
+		ctx->diff.invalid_box = SV_BORDER_BOX;
+	}
+}
+
+static void Widget_OnRefreshCanvas(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
+{
+	if (ctx->diff.invalid_box < SV_GRAPH_BOX) {
+		ctx->diff.invalid_box = SV_GRAPH_BOX;
+	}
+}
+
+static void Widget_ComputeFlexLayoutStyle(LCUI_Widget w)
+{
+	LCUI_FlexLayoutStyle *data = &w->computed_style.flex;
+	LCUI_Style s = StyleSheet_GetStyle(w->style, key_justify_content);
+
+	if (s->type != LCUI_STYPE_STYLE || !s->is_valid) {
+		data->justify_content = SV_FLEX_START;
+		return;
+	}
+	data->justify_content = s->val_style;
+}
+
+static void Widget_UpdateLayout(LCUI_Widget w)
+{
+	if (w->computed_style.display == SV_FLEX) {
+		Widget_ComputeFlexLayoutStyle(w);
+	}
+	Widget_AddTask(w, LCUI_WTASK_LAYOUT);
+}
+
+static void Widget_OnUpdateBorder(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
+{
+	const LCUI_BorderStyle *old_style = &ctx->diff.border;
+	const LCUI_BorderStyle *new_style = &w->computed_style.border;
+
+	Widget_ComputeBorderStyle(w);
+	/* 如果边框变化导致图层尺寸变化 */
+	if (old_style->top.width != new_style->top.width ||
+	    old_style->right.width != new_style->right.width ||
+	    old_style->bottom.width != new_style->bottom.width ||
+	    old_style->left.width != new_style->left.width) {
+		Widget_UpdateLayout(w);
+		Widget_AddTask(w, LCUI_WTASK_POSITION);
+	}
+}
+
+static void Widget_OnUpdateVisible(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
+{
+	LCUI_Style s = &w->style->sheet[key_visibility];
+	LCUI_BOOL visible = w->computed_style.visible;
+
+	if (w->computed_style.display == SV_NONE) {
+		w->computed_style.visible = FALSE;
+	} else if (s->is_valid && s->type == LCUI_STYPE_STRING &&
+		   strcmp(s->val_string, "hidden") == 0) {
+		w->computed_style.visible = FALSE;
+	} else {
+		w->computed_style.visible = TRUE;
+	}
+}
+
+static void Widget_OnUpdateDisplay(LCUI_Widget w, LCUI_WidgetTaskContext ctx)
+{
+	LCUI_Style s = &w->style->sheet[key_display];
+	LCUI_WidgetStyle *style = &w->computed_style;
+
+	if (s->is_valid && s->type == LCUI_STYPE_STYLE) {
+		style->display = s->style;
+	} else {
+		style->display = SV_BLOCK;
+	}
+	if (style->display == SV_NONE) {
+		w->computed_style.visible = FALSE;
+	}
```

Comment:
if (s->is_valid && s->type == LCUI_STYPE_STYLE) {
	style->display = s->style;

	// SV_NONE value maybe only in this case
	if (style->display == SV_NONE) {
		w->computed_style.visible = FALSE;
	}
} else {
	style->display = SV_BLOCK;
}